lib/src/default_index/revset_engine.rs

```diff
@@ -130,8 +130,13 @@ impl<I: AsCompositeIndex> Revset for RevsetImpl<I> {
         self.entries().next().is_none()
     }
 
-    fn count(&self) -> usize {
-        self.entries().count()
+    fn count_estimate(&self) -> (usize, Option<usize>) {
```

Comment:
> My concern is that I wouldn't expect size_hint to iterate 50 times (like what I'm doing now) or iterate through the whole revset (as Martin suggests below), 

I think iterating constant times is okay. If we do the latter, I prefer not using `size_hint()`. `count_estimate()` or `count()` sounds better.

Reply:
I'm fine with either. IIUC, @yuja prefers iterating a constant number of times and prefers to call the function `size_hint()` if we do?

Reply:
Nah, it's just about naming. For the existing consumers, I think it's simpler to iterate the whole revset, and I wouldn't call such method as `size_hint()`.

Reply:
@yuja , regarding iterating the whole revset, what do you think of my justification of only iterating part of the revset (see the comment on revset_engine.rs, starting with "Even on local disk, GitHub")?

Reply:
How slow is it in some real-life repos? In my linux-stable repo with 100 branches, of which one is ~1.1M commits behind, `jj branch list` takes about 1.1s (0.9s if no branches are behind). In the same situation, git `git status` takes 12.6s (I suppose I don't have that optimization from yet). So unless there are other cases where it's unbearably slow, I don't see much reason to not return the exact count. I don't really mind it much either, but I prefer to have a concrete example of where it's slow before we decide to return an approximation.

Reply:
Here is how I see things:

1. In order to support "branch list" with non-fully-local backends, "branch list" needs to be able to report approximate ahead/behind.
2. And we want to test that too.
3. In order for "branch list" to be able to report approximate ahead/behind, Revset needs to be able to report approximations.
4. If we change default_index's Revset to report approximations, it will allow us to test "branch list". Maybe the effect on the end user will be slightly negative - they will have a change in UX and only see approximate numbers, and in return, they will only get a very small increase in speed. (Well, maybe they will feel psychologically better because the computer didn't do unnecessary work to give them a number they didn't need.) Maybe there will be more benefits visible to the end user if we have commands calculating ahead/behind in bulk, like what GitHub did, but this probably won't happen in the short term. (My original argument of execution time reduction was from GitHub's effort to speed this up, but looks like this isn't true in our case.)

And here are the ways in which we can proceed:

1. Drop this PR and figure out another way to support "branch list" with non-fully-local backends (e.g. have each backend carry a flag saying whether revision walking is cheap or expensive, and/or have each backend have a custom way of reporting ahead/behind - but we will run into the same problem of testing).
2. Implement the API change like in this PR, but make the default_index Revset always iterate and thus always return exact values. This means that "branch list" cannot be tested (and thus, anyone implementing a Revset that estimates counts will need to, say, implement their own test to make sure that things don't break).
3. Implement what's in this PR. (Maybe rename count_estimate to size_hint.)

Or maybe there's something between 2 and 3 that I can't think of.

I still think we should do 3 for the reason of testing, but I do acknowledge that it seems like we're contorting the design too much - we are exercising a code path in production mainly so that we can make sure it works (in a unit test).

Reply:
Another option is to define a crate feature (conditional compilation) so the implementation can limit the walk to 50 (or even 10) commits in tests but not in normal builds. We have this `test-fakes` feature defined [here](https://github.com/martinvonz/jj/blob/b0cc64b9a3aedaeb8eeafd74889ca209f6c62864/cli/Cargo.toml#L81). We could use that (possibly renaming it). Then the `count_estimate()` function would look something like this:

```rust
if cfg!(test-fakes) {
  let count = self.entries().take(50).count();
  if count < 50 {
    (count, Some(count))
  } else {
    (50, None)
  }
} else {
  self.iter().count()
}
```

Reply:
I agree with @martinvonz here. Scanning the index wouldn't be slow compared to Git object traversal, and it's nice if we can avoid choosing the magic number `50`.

Regarding tests, I think it's okay to go without an e2e test. That's not ideal, but the current implementation looks readable and less likely to get broken. Anyway, it's undesirable to execute "jj new" ~100 times in cli tests because the cli tests are slow.

Reply:
OK, done. We might want to discuss the name but if we decide on something different, that should not be too difficult to change.

Reply:
Going with count_estimate and testing.